overlord/snapstate/snapstate.go

```diff
@@ -1358,6 +1358,102 @@ func installWithDeviceContext(st *state.State, name string, opts *RevisionOption
 	return doInstall(st, &snapst, snapsup, 0, fromChange, nil)
 }
 
+func Download(ctx context.Context, st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext) (*state.TaskSet, error) {
+	return download(ctx, st, name, opts, userID, flags, deviceCtx)
+}
+
+func download(ctx context.Context, st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext) (*state.TaskSet, error) {
+	if opts == nil {
+		opts = &RevisionOptions{}
+	}
+
+	if opts.CohortKey != "" && !opts.Revision.Unset() {
+		return nil, errors.New("cannot specify revision and cohort")
+	}
+
+	if opts.Channel == "" {
+		opts.Channel = "stable"
+	}
+
+	var snapst SnapState
+	err := Get(st, name, &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	// if the caller does not specify a revision and the snap is already
+	// installed, then there is nothing to do. also, if the caller does specify
+	// a revision and the revision is already installed, then there is nothing
+	// to do.
+	if (opts.Revision.Unset() || opts.Revision == snapst.Current) && snapst.IsInstalled() {
+		return nil, &snap.AlreadyInstalledError{Snap: name}
+	}
+
+	if err := snap.ValidateInstanceName(name); err != nil {
+		return nil, fmt.Errorf("invalid instance name: %v", err)
+	}
+
+	sar, err := downloadInfo(ctx, st, name, opts, userID, deviceCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	info := sar.Info
+
+	if flags.RequireTypeBase && info.Type() != snap.TypeBase && info.Type() != snap.TypeOS {
+		return nil, fmt.Errorf("unexpected snap type %q, instead of 'base'", info.Type())
+	}
+
+	// TODO: should changeKind be "download"? i don't see that as one of the
+	// options that this function operates on
+	if err := checkDiskSpace(st, "install", []minimalInstallInfo{installSnapInfo{info}}, userID); err != nil {
+		return nil, err
+	}
+
+	providerContentAttrs := defaultProviderContentAttrs(st, info)
+	snapsup := &SnapSetup{
+		Channel:            opts.Channel,
+		Base:               info.Base,
+		Prereq:             getKeys(providerContentAttrs),
+		PrereqContentAttrs: providerContentAttrs,
+		UserID:             userID,
+		Flags:              flags.ForSnapSetup(),
+		DownloadInfo:       &info.DownloadInfo,
+		SideInfo:           &info.SideInfo,
+		Type:               info.Type(),
+		Version:            info.Version,
+		PlugsOnly:          len(info.Slots) == 0,
+		InstanceKey:        info.InstanceKey,
+		auxStoreInfo: auxStoreInfo{
+			Media: info.Media,
+			// TODO: still needed here?
+			// XXX we store this for the benefit of old snapd
+			Website: info.Website(),
+		},
+		CohortKey:          opts.CohortKey,
+		ExpectedProvenance: info.SnapProvenance,
```

Comment:
I think we should not set things not used by the subset of tasks we have here, for example I don't think we need Prereq or PrereqContentAttrs, PlugsOnly or the auxStoreInfo?

Reply:
Makes sense, I grabbed this from another test.